### PR TITLE
fix: change list check

### LINKED ISSUE #69: Change output list should be verified to avoid breaking the UI
## Description

If the change list has duplicated outputs, meaning 2 entries saying that output index X is a change, we will break the output confirming UI.
For instance, if we have 3 outputs and 2 entries on the change list saying that output index 0 is change we will show to the user to confirm outputs:

- `Output 1/1`
- `Output 2/1`

This can be specially bad if we had 2 outputs, since it actually shows `Output 1/0`, it can also show `Output 1/-1` depending on how we manipulate this.

## Solution

When parsing the change list we need to check for duplicates and fail the transaction if any are found.
Since we do not have data on the transaction yet we can parse the change list and when we parse the number of outputs on the transaction we also need to check that the highest change index is on the transaction (since we do not have duplicates this also checks that the number of change outputs is lower than the number of outputs).

### PR DESCRIPTION
**Pull request by r4mmer**
## Acceptance criteria

- Fix issue described in #69 
- Clean global context when an error is returned that ends the sign_tx call.

Due to the new edge case tests I found out another issue, some errors would return an error code but would not clean the device state.
This is somewhat complicated since it would fail the next call even if it has nothing to do with the error.
To fix this we need to clean the global context before throwing, the best way to do this is concentrate the throws in a higher order method and change some methods to return an error code instead of THROWing it.

### CODE REVIEW

File: src/handler/sign_tx.c [hunk 1]
```diff
@@ -65,29 +65,46 @@ bool verify_change(change_output_info_t *info, tx_output_t output) {
  *
  * The outputs will be on the global context for sign tx (`tx_info`)
  **/
-void read_change_info_v1(buffer_t *cdata) {
+uint16_t read_change_info_v1(buffer_t *cdata) {
```
Comment (r4mmer) [inline src/handler/sign_tx.c:68]: We are changing these methods to return an error code instead of throwing one because this method is called from `receive_data` which is outside the `decode_elements` try..catch.

This means that the error is raised to the try..catch of the main function and just returns the error code to the user without cleaning the global context.

The reason this needs to change is because the next call will reach the device while the global context is initialized, and depending on the call it may return a bad state error (even on a perfectly correct call).

File: src/handler/sign_tx.c [hunk 2]
```diff
@@ -65,29 +65,46 @@ bool verify_change(change_output_info_t *info, tx_output_t output) {
...
         change_output_info_t *info = &G_context.tx_info.change_info[i];
 
         // 1 byte for change output index
         if (!buffer_read_u8(cdata, &info->index)) {
-            THROW(SW_WRONG_DATA_LENGTH);
+            return SW_WRONG_DATA_LENGTH;
         }
 
+        // check for duplicates
+        for (uint8_t j = 0; j < indexes_len; j++) {
```
Comment (r4mmer) [inline src/handler/sign_tx.c:91]: Since indexes_len can only go up to `TX_MAX_TOKENS + 1` (which is 11) we don't have to enhance this search.

File: src/handler/sign_tx.c [hunk 3]
```diff
@@ -474,13 +515,23 @@ bool receive_data(buffer_t *cdata, uint8_t chunk) {
         }
 
         init_sign_tx_ctx();
+        uint16_t error_code = 0;
         // read change output and sighash
-        read_change_info(cdata);
+        error_code = read_change_info(cdata);
+        if (error_code != 0) {
+            explicit_bzero(&G_context, sizeof(G_context));
+            THROW(error_code);
```
Comment (r4mmer) [inline src/handler/sign_tx.c:523]: Not cleaning the state was making the next call to the Ledger fail due to bad state.
It was very unlikely to throw an error here since the transaction would need to be extremely malformed.
But since we added stricter rules for the change list we may reach this error more often, hence the need to make this context cleanup.

File: tests/test_sign_cmd.py [hunk 4]
```diff
@@ -32,7 +32,7 @@ def test_sign_tx_change_old_protocol(cmd, public_key_bytes):
         )
         for x in range(5)
     ]
-    change_index = fake.pyint(0, 5)
+    change_index = fake.pyint(0, 4)
```
Comment (r4mmer) [inline tests/test_sign_cmd.py:35]: `fake.pyint` generates from 0 to 5 inclusive, this had a 1/6 chance of failing if we chose the 5 as an index.

File: src/handler/sign_tx.c (hunk 1 above)
Comment (tuliomir) [inline src/handler/sign_tx.c:68]: chore: We need to update the docstrings with the return value and its meaning

File: src/handler/sign_tx.c [hunk 5]
```diff
@@ -101,26 +118,26 @@ void read_change_info_v1(buffer_t *cdata) {
  *
  * The outputs will be on the global context for sign tx (`tx_info`)
  **/
-void read_change_info_old_protocol(uint8_t change_byte, buffer_t *cdata) {
+uint16_t read_change_info_old_protocol(uint8_t change_byte, buffer_t *cdata) {
```
Comment (tuliomir) [inline src/handler/sign_tx.c:121]: chore: Same docstrings update needed here

File: src/handler/sign_tx.c [hunk 6]
```diff
@@ -141,10 +160,10 @@ void read_change_info_old_protocol(uint8_t change_byte, buffer_t *cdata) {
  * First byte is the version byte for the protocol (for transactions)
  * It will be used to differentiate the old and new protocols
  **/
-void read_change_info(buffer_t *cdata) {
+uint16_t read_change_info(buffer_t *cdata) {
```
Comment (tuliomir) [inline src/handler/sign_tx.c:163]: chore: Same docstrings update needed here

File: src/handler/sign_tx.c [hunk 7]
```diff
@@ -187,7 +209,7 @@ void read_change_info(buffer_t *cdata) {
  *
  * The output will be on the global context for sign tx (`tx_info`)
  **/
-void read_tx_data(buffer_t *cdata) {
+uint16_t read_tx_data(buffer_t *cdata) {
```
Comment (tuliomir) [inline src/handler/sign_tx.c:212]: chore: Same docstrings update needed here

Comment (r4mmer) [reply src/handler/sign_tx.c:220]: Done

File: src/handler/sign_tx.c (hunk 1 above)
Comment (r4mmer) [reply src/handler/sign_tx.c:70]: Done

File: src/handler/sign_tx.c (hunk 5 above)
Comment (r4mmer) [reply src/handler/sign_tx.c:125]: Done

File: src/handler/sign_tx.c (hunk 6 above)
Comment (r4mmer) [reply src/handler/sign_tx.c:169]: Done

**Review by tuliomir — APPROVED** [review]

**Review by pedroferreira1 — APPROVED** [review]

File: src/handler/sign_tx.c [hunk 8]
```diff
@@ -197,8 +227,27 @@ void read_tx_data(buffer_t *cdata) {
           buffer_read_u8(cdata, &G_context.tx_info.remaining_inputs) &&
           buffer_read_u8(cdata, &G_context.tx_info.outputs_len))) {
         // if an error occurs reading
-        THROW(SW_WRONG_DATA_LENGTH);
+        return SW_WRONG_DATA_LENGTH;
+    }
+
+    for (uint8_t i = 0; i < G_context.tx_info.change_len; i++) {
+        // check that all change indexes are inside the outputs array
+        if (G_context.tx_info.change_info[i].index >= G_context.tx_info.outputs_len) {
```
Comment (msbrogli) [inline src/handler/sign_tx.c:235]: I'd like to confirm that we confirm that all change outputs have addresses that belong to this wallet.

**Review by msbrogli — APPROVED** [review]

File: src/handler/sign_tx.c (hunk 8 above)
Comment (r4mmer) [reply src/handler/sign_tx.c:235]: We have a method `verify_change` (under `src/handler/sign_tx.c)` that checks this and runs for all change outputs on the transaction.

Comment (msbrogli) [reply src/handler/sign_tx.c:235]: Thanks!
